homeassistant/components/zhong_hong/climate.py

```diff
@@ -2,138 +2,65 @@
-        vol.Optional(CONF_PORT, default=DEFAULT_PORT): cv.port,
-        vol.Optional(
-            CONF_GATEWAY_ADDRRESS, default=DEFAULT_GATEWAY_ADDRRESS
-        ): cv.positive_int,
-    }
-)
-
-ZHONG_HONG_MODE_COOL = "cool"
-ZHONG_HONG_MODE_HEAT = "heat"
-ZHONG_HONG_MODE_DRY = "dry"
-ZHONG_HONG_MODE_FAN_ONLY = "fan_only"
-
-
+# Mode mapping
 MODE_TO_STATE = {
     ZHONG_HONG_MODE_COOL: HVACMode.COOL,
     ZHONG_HONG_MODE_HEAT: HVACMode.HEAT,
     ZHONG_HONG_MODE_DRY: HVACMode.DRY,
     ZHONG_HONG_MODE_FAN_ONLY: HVACMode.FAN_ONLY,
 }
 
-# HA → zhong_hong
-FAN_MODE_MAP = {
-    FAN_LOW: "LOW",
-    FAN_MEDIUM: "MID",
-    FAN_HIGH: "HIGH",
-    FAN_MIDDLE: "MID",
-    "medium_high": "MIDHIGH",
-    "medium_low": "MIDLOW",
-}
-FAN_MODE_REVERSE_MAP = {v: k for k, v in FAN_MODE_MAP.items()}
-
+STATE_TO_MODE = {v: k for k, v in MODE_TO_STATE.items()}
 
-def setup_platform(
+async def async_setup_entry(
     hass: HomeAssistant,
-    config: ConfigType,
-    add_entities: AddEntitiesCallback,
-    discovery_info: DiscoveryInfoType | None = None,
+    entry: ZhongHongConfigEntry,
+    async_add_entities: AddConfigEntryEntitiesCallback,
 ) -> None:
-    """Set up the ZhongHong HVAC platform."""
-
-    host = config.get(CONF_HOST)
-    port = config.get(CONF_PORT)
-    gw_addr = config.get(CONF_GATEWAY_ADDRRESS)
-    hub = ZhongHongGateway(host, port, gw_addr)
-    devices = [
-        ZhongHongClimate(hub, addr_out, addr_in)
-        for (addr_out, addr_in) in hub.discovery_ac()
-    ]
-
-    _LOGGER.debug("We got %s zhong_hong climate devices", len(devices))
-
-    hub_is_initialized = False
-
-    def _start_hub():
-        """Start the hub socket and query status of all devices."""
-        hub.start_listen()
-        hub.query_all_status()
-
-    async def startup():
-        """Start hub socket after all climate entity is set up."""
-        nonlocal hub_is_initialized
-        if not all(device.is_initialized for device in devices):
-            return
+    """Set up ZhongHong climate entities."""
+    coordinator = entry.runtime_data
 
-        if hub_is_initialized:
-            return
-
-        _LOGGER.debug("zhong_hong hub start listen event")
-        await hass.async_add_executor_job(_start_hub)
-        hub_is_initialized = True
-
-    async_dispatcher_connect(hass, SIGNAL_DEVICE_ADDED, startup)
-
-    # add devices after SIGNAL_DEVICE_SETTED_UP event is listened
-    add_entities(devices)
-
-    def stop_listen(event):
-        """Stop ZhongHongHub socket."""
-        hub.stop_listen()
+    entities = [
+        ZhongHongClimate(coordinator, device_id, device)
+        for device_id, device in coordinator.devices.items()
+    ]
 
-    hass.bus.listen_once(EVENT_HOMEASSISTANT_STOP, stop_listen)
+    async_add_entities(entities)
 
 
-class ZhongHongClimate(ClimateEntity):
-    """Representation of a ZhongHong controller support HVAC."""
+class ZhongHongClimate(
+    CoordinatorEntity[ZhongHongDataUpdateCoordinator], ClimateEntity
+):
```

Comment:
ideally we do the coordinator change in a separate PR to keep this PR well-scoped